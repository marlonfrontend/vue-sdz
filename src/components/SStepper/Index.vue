<template>
  <div class="s-stepper">
    <div class="stepper-tracker" />

    <ul class="stepper">
      <li
        v-for="(item, stepIndex) in items"

        :key="stepIndex"
        :class="getClasses(item, stepIndex)"

        @click="$emit('select', item)"
      >
        <div class="content">
          <label>{{ item.label }}</label>

          <s-icon v-if="item.icon" :icon="item.icon" v-bind="$attrs" />
        </div>
      </li>
    </ul>
  </div>
</template>

<script>
export default {
  name: 'SStepper',

  components: {
    SIcon: () => import('../SIcon/Index.vue').then(c => c.default)
  },

  props: {
    step: Number,

    items: { type: Array, required: true }
  },

  data () {
    return {
      progressiveStep: 1
    }
  },

  watch: {
    step: 'incrementStep',

    progressiveStep: {
      handler: 'incrementStep',
      immediate: true
    }
  },

  methods: {
    incrementStep () {
      setTimeout(() => {
        if (this.progressiveStep <= this.step) this.progressiveStep++
      }, 1000)
    },

    getClasses (item, stepIndex) {
      const isActive = stepIndex <= (this.progressiveStep - 1)

      return ['step', {
        '--is-active': isActive,
        '--is-disabled': item.disabled,
        '--is-current-step': isActive && this.step === stepIndex
      }]
    }
  }
}
</script>

<style lang="scss">
@import "./src/styles/_index.scss";

.s-stepper {
  & > .stepper-tracker {
    position: relative;

    &::after {
      content: '';
      z-index: -2;

      height: 2px;
      width: calc(100% - 20%);

      top: 15px;
      left: 10%;
      position: absolute;

      background: color(base, light);
    }
  }

  & > .stepper {
    @keyframes bounce {
      from { width: 0%; }
      to { width: 100%; }
    }

    margin: 0;
    width: 100%;
    display: flex;
    counter-reset: step;
    align-items: baseline;
    justify-content: space-between;

    & > .step {
      width: 100%;
      font-size: 12px;
      position: relative;
      text-align: center;
      list-style-type: none;

      & > .content {
        display: flex;
        align-items: center;
        justify-content: center;

        & > label {
          font-size: 12px;
          transition: opacity .7s ease;
        }
      }

      &::before {
        width: 30px;
        height: 30px;

        display: block;

        content: counter(step);
        counter-increment: step;

        display: flex;
        align-items: center;
        justify-content: center;

        border-width: 2px;
        border-radius: 50%;
        border-style: solid;

        margin: 0 auto -55px auto;
        background-color: color(neutral, base);
        transition: border-color .3s ease-in-out;
      }

      &:first-child:after { content: none; }
    }

    .--is-active {
      &::before {
        color: color(primary, base);
        border-color: color(primary, base);
      }

      &::after {
        content: '';
        z-index: -1;

        width: 100%;
        height: 2px;

        top: 15px;
        left: -50%;
        position: absolute;

        background-color: color(primary, base);
        transition: background-color .3s ease-in-out;

        animation: bounce 0.7s;
      }

      & > .content:last-child { color: color(primary, base); }
    }

    .--is-disabled {
      cursor: not-allowed;

      & > .content { color: gray !important; }

      &::before {
        // color: color(neutral, light) !important;
        // border-color: color(neutral, light) !important;
        color: gray !important;
        border-color: gray !important;
      }
    }
    .--is-current-step > .content { font-weight: 800; }
  }
}
</style>
