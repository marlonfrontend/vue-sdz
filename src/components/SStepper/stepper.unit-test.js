import { mount } from '@vue/test-utils'

import SStepper from './Index.vue'

/* eslint-disable */
describe('SStepper', () => {
  let cmp;

  beforeEach(() => {
    cmp = mount(SStepper, {
      propsData:{
        step: 2,
        items: [
          { label: 'item 1', icon: 'sdz-dollar-sign', disabled: false },
          { label: 'item 2', icon: 'sdz-link', disabled: false },
          { label: 'item 3', icon: 'sdz-shield', disabled: false },
          { label: 'item 4', icon: 'sdz-shield-off', disabled: true },
          { label: 'item 5', icon: 'sdz-cart', disabled: false }
        ]
      }
    });
  });


  test('is instantiated', () => {
    expect(cmp.exists()).toBeTruthy()
  })
  

  test("has the expected html structure", () => {
    expect(cmp.element).toMatchSnapshot();
  })

  test('count items', () => {

    const items = cmp.findAll('.content')

    expect(items.length).toBe(5)
  })

  test('check item', () => {

    const cl = cmp.find('.--is-active > div > label')

    expect(cl.text()).toContain('item 1')
  })

})
