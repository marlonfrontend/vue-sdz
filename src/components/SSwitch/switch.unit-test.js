import { mount } from '@vue/test-utils'

import SStepper from './Index.vue'

/* eslint-disable */
describe('SStepper', () => {
  let cmp;

  beforeEach(() => {
    cmp = mount(SStepper, {
      propsData:{
        value: true,
        label: 'Default Switch',
        negative: true
      }
    });
  });


  test('is instantiated', () => {
    expect(cmp.exists()).toBeTruthy()
  })
  

  test("has the expected html structure", () => {
    expect(cmp.element).toMatchSnapshot();
  })

  test('check label text', () => {

    const span = cmp.find('.text')
    expect(span.exists()).toBe(true)
    expect(span.text()).toBe('Default Switch')

  })


  test('check is checked', () => {

    expect(cmp.classes()).toContain('s-switch')

  })

  test("check setChecked false", async() => {

    expect(cmp.classes()).toContain('s-switch')

    const check = await cmp.find('input[type="checkbox"]')
    
    check.setChecked(false)
    expect(check.element.checked).toBe(false)
  })

})
