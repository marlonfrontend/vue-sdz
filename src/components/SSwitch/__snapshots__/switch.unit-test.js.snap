// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SStepper has the expected html structure 1`] = `
<label
  class="s-switch --is-checked --is-negative"
>
  <input
    class="input"
    type="checkbox"
  />
   
  <span
    class="switch"
  />
   
  <span
    class="text"
  >
    Default Switch
  </span>
</label>
`;
